src/df/df.rs

```diff
@@ -0,0 +1,637 @@
+static OPT_HUMAN_READABLE_2: &str = "human-readable-2";
+static OPT_INODES: &str = "inodes";
+static OPT_KILO: &str = "kilo";
+static OPT_LOCAL: &str = "local";
+static OPT_NO_SYNC: &str = "no-sync";
+static OPT_OUTPUT: &str = "output";
+static OPT_PORTABILITY: &str = "portability";
+static OPT_SYNC: &str = "sync";
+static OPT_TYPE: &str = "type";
+static OPT_PRINT_TYPE: &str = "print-type";
+static OPT_EXCLUDE_TYPE: &str = "exclude-type";
+static OPT_VERSION: &str = "version";
+
+static MOUNT_OPT_BIND: &str = "bind";
+
+#[cfg(any(target_os = "freebsd", target_os = "macos"))]
+#[repr(C)]
+#[derive(Copy, Clone)]
+#[allow(non_camel_case_types)]
+struct fsid_t {
+    val: [i32; 2usize],
+}
+
+#[cfg(all(target_os = "macos", target_arch = "x86_64"))]
+#[repr(C)]
+#[derive(Copy, Clone)]
+#[allow(non_camel_case_types)]
+struct statfs {
+    f_bsize: u32,
+    f_iosize: i32,
+    f_blocks: u64,
+    f_bfree: u64,
+    f_bavail: u64,
+    f_files: u64,
+    f_ffree: u64,
+    f_fsid: fsid_t,
+    f_owner: uid_t,
+    f_type: u32,
+    f_flags: u32,
+    f_fssubtype: u32,
+    f_fstypename: [c_char; 16usize],
+    f_mntonname: [c_char; 1024usize],
+    f_mntfromname: [c_char; 1024usize],
+    f_reserved: [u32; 8usize],
+}
+
+#[cfg(all(target_os = "freebsd", not(all(target_os = "macos", target_arch = "x86_64"))))]
+#[repr(C)]
+#[derive(Copy, Clone)]
+#[allow(non_camel_case_types)]
+struct statfs {
+    f_version: u32,
+    f_type: u32,
+    f_flags: u64,
+    f_bsize: u64,
+    f_iosize: u64,
+    f_blocks: u64,
+    f_bfree: u64,
+    f_bavail: i64,
+    f_files: u64,
+    f_ffree: i64,
+    f_syncwrites: u64,
+    f_asyncwrites: u64,
+    f_syncreads: u64,
+    f_asyncreads: u64,
+    f_spare: [u64; 10usize],
+    f_namemax: u32,
+    f_owner: uid_t,
+    f_fsid: fsid_t,
+    f_charspare: [c_char; 80usize],
+    f_fstypename: [c_char; 16usize],
+    f_mntfromname: [c_char; 88usize],
+    f_mntonname: [c_char; 88usize],
+}
+
+
+/// Store names of file systems as a selector.
+/// Note: `exclude` takes priority over `include`.
+struct FsSelector {
+    include: HashSet<String>,
+    exclude: HashSet<String>,
+}
+
+struct Options {
+    show_local_fs: bool,
+    show_all_fs: bool,
+    show_listed_fs: bool,
+    show_fs_type: bool,
+    show_inode_instead: bool,
+    print_grand_total: bool,
+    block_size: i64,
+    human_readable_base: i64,
+    fs_selector: FsSelector,
+}
+
+#[derive(Debug, Clone)]
+struct MountInfo {
+    dev_id: i32,
+    dev_name: String,
+    fs_type: String,
+    mount_dir: String,
+    mount_option: String, // we only care "bind" option
+    mount_root: String,
+    remote: bool,
+    dummy: bool,
+}
+
+#[derive(Debug, Clone)]
+struct FsUsage {
+    blocksize: u64,
+    blocks: u64,
+    bfree: u64,
+    bavail: u64,
+    bavail_top_bit_set: bool,
+    files: u64,
+    ffree: u64
+}
+
+#[derive(Debug, Clone)]
+struct Filesystem {
+    mountinfo: MountInfo,
+    usage: FsUsage,
+}
+
+macro_rules! exit_with_error(
```

Comment:
I believe this does the same thing as the `crash!` macro.

Reply:
yep, I didn't know that...